web_monitoring/db.py

```diff
@@ -172,14 +181,31 @@ def from_env(cls):
    WEB_MONITORING_DB_URL (optional)
    WEB_MONITORING_DB_EMAIL
    WEB_MONITORING_DB_PASSWORD
+   WEB_MONITORING_DB_TIMEOUT (optional)
 
 Alternatively, you can instaniate Client(user, password) directly.""")
-        return cls(email=email, password=password, url=url)
+        
+        if timeout_string is None:
+            return cls(email=email, password=password, url=url)
+
+        try:
+            timeout = float(timeout_string)
+            return cls(email=email, password=password, url=url, timeout=timeout)
+        except ValueError:
+            raise ValueError('Unable to parse WEB_MONITORING_DB_TIMEOUT as a float, '
+                             f'{timeout_string} provided')
 
-    def request(self, method, url, data=None, **kwargs):
+    def request(self, method, url, data=None, timeout=None, **kwargs):
         if not url.startswith('http://') and not url.startswith('https://'):
             url = f'{self._api_url}{url}'
 
+        if timeout is not None and timeout < 0:
+            raise ValueError(f'Timeout must be non-negative. "{timeout}" was provided.')
+        elif timeout == 0:
+            timeout = None
+        elif timeout is None:
+            timeout = self._timeout
+
```

Comment:
I don't love this block of code but couldn't think of anything better.

Reply:
I love the idea of abstracting it out into a separate function.

If we want to support 'no timeout' as a state, I think 0 is a good value for it. The `NO_VALUE = object()` reads really weirdly to me (although I'm not an expert). I'll let you decide. Same for whether we validate, urllib3 actually has pretty good error messages:
```
ValueError: Attempted to set connect timeout to -1, but the timeout cannot be set to a value less than or equal to 0
```

Reply:
> The `NO_VALUE = object()` reads really weirdly to me

Oh, for sure. You won’t see this too often, and it’s generally never meant to be a user-visible thing, but it’s a relatively common practice if you need to be able to distinguish between something being set to `None` and not being set at all. `object()` is the simplest way to get a unique value that will never be equal to anything else in Python.

We do something similar for our `FiniteQueue` implementation, but use a class so that we get nice names when debugging (because, unlike `NO_VALUE`, that one is likely to be more visible to users of the code): https://github.com/edgi-govdata-archiving/web-monitoring-processing/blob/93e9e7cb80ef85e973717df46f270c72cbcf1022/web_monitoring/utils.py#L167-L169

**Anyway! I am cool if you want `0` to represent “no timeout” and `None` to represent “I didn’t specify” so you don’t need to do the above. :)**

Reply:
Ok that actually makes a lot of sense, thanks for the example. Looking at it kind of reminded me of how Scala uses a `Nil` singleton to represent empty lists. Here you're just defining your own singleton to represent `a value wasn't set`.

If you're still happy with `0`, I'm still happy with `0` :)

Reply:
Yep, all good. 👍